### PR TITLE
Feature/registration

### PR DESCRIPTION
**Pull request by Mykyta-Pysarev**
It`s working container. Some additional design features could be added later.

### CODE REVIEW

**Review by KirillSerg — APPROVED** [review]

**Review by KirillSerg — APPROVED** [review]
looks good for me

File: src/containers/Registration.tsx [hunk 1]
```diff
@@ -0,0 +1,209 @@
+import { yupResolver } from "@hookform/resolvers/yup";
+import React, { useState } from "react";
+import { Modal, Button, Form, FloatingLabel } from "react-bootstrap";
+import "bootstrap/dist/css/bootstrap.css";
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]:
```suggestion
```

File: src/containers/Registration.tsx [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  confirmPassword: yup
+    .string()
+    .oneOf([yup.ref("password"), null], "Пароль має співпадати"),
+});
+
+const Registration = () => {
+  const [show, setShow] = useState(false);
+
+  const handleClose = () => setShow(false);
+  const handleShow = () => setShow(true);
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: can be replaced with single `toggleShowState()` function

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

File: src/containers/Registration.tsx [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import { useForm } from "react-hook-form";
+import styled from "styled-components";
+import * as yup from "yup";
+
+const WrapperDiv = styled.div`
+  > * {
+    font-family: TwitterChirp, -apple-system, BlinkMacSystemFont, "Segoe UI",
+      Roboto, Helvetica, Arial, sans-serif;
+  }
+`;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: thanks, but let's drop it. It'll be added in another place

File: src/containers/Registration.tsx [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  padding-right: 0;
+  margin-right: 0;
+  min-width: 124%;
+  transform-origin: 0 0;
+`;
+
+interface MyForm {
+  userName: string;
+  email: string;
+  dateOfBirth: Date;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: not sure if it's proper type
common practice is to use ISO string data format to save dates
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString

Comment (alexandrtovmach) [reply src/containers/Registration.tsx]: ISO should be a string type

File: src/containers/Registration.tsx [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        Зареєструватися
+      </Button>
+
+      <Modal size="lg" centered show={show} onHide={handleClose}>
+        <Modal.Header closeButton style={{ margin: "0px" }}>
+          {" "}
+        </Modal.Header>
+        <StyledModalTitle>Створіть свій профіль</StyledModalTitle>
+        <Modal.Body>
+          <Form
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: move form to separate container

File: src/containers/Registration.tsx [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            style={{
+              paddingLeft: "2rem",
+              paddingRight: "2rem",
+            }}
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: no inline styles

Comment (alexandrtovmach) [reply src/containers/Registration.tsx]: move it to styled-components

File: src/containers/Registration.tsx [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+            <StyledFloatingLabel controlId="userNameInput" label="Ім'я">
+              <StyledFormControl
+                {...register("userName")}
+                type="text"
+                name="userName"
+                placeholder="Ім'я"
+                maxLength={50}
+              />
+              <StyledCounterLabel>
+                {watch("userName") ? watch("userName").length : 0} / 50
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: add variable before return to use only one watch() call

File: src/pages/Home.tsx [hunk 8]
```diff
@@ -1,13 +1,17 @@
-import FormValidationSample from "../components/FormValidationSample";
+// import FormValidationSample from "../components/FormValidationSample";
```
Comment (alexandrtovmach) [inline src/pages/Home.tsx]: remove comment

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

**Comment by ArtemLigerko** [comment]
Propose clear registration forms after closing it, now it save personal data

File: src/containers/Registration.tsx (hunk 1 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 2 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 3 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Deleted

File: src/containers/Registration.tsx (hunk 4 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Change date type to string

File: src/containers/Registration.tsx (hunk 7 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Done

File: src/pages/Home.tsx (hunk 8 above)
Comment (Mykyta-Pysarev) [reply src/pages/Home.tsx]: Deleted

File: src/containers/Registration.tsx (hunk 6 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 5 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Done

**Comment by Mykyta-Pysarev** [comment]
image

File: package.json [hunk 9]
```diff
@@ -9,6 +9,10 @@
     {
       "name": "Artem Ligerko",
       "email": "ligerko.aa@gmail.com"
+    },
+    {
+      "name": "Mykyta Pysarev",
```
Comment (KirillSerg) [inline package.json:14]: )

**Review by KirillSerg — APPROVED** [review]

File: src/containers/Registration.tsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+interface MyForm {
+  userName: string;
+  email: string;
+  dateOfBirth: string;
+  password: string;
+  confirmPassword: string;
+}
+
+const schema = yup.object().shape({
+  userName: yup.string().required("Як вас звати?"),
```
Comment (KirillSerg) [inline src/containers/Registration.tsx]: maybe move the signature to the right

File: src/containers/ModalAuthorization.tsx [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  font-size: 2rem;
+`;
+
+const ModalAuthorization = () => {
+  const [show, setShow] = useState(false);
+
+  const toggleShowState = () => setShow(!show);
+  return (
+    <>
+      <Button variant="primary" onClick={toggleShowState}>
```
Comment (KirillSerg) [inline src/containers/ModalAuthorization.tsx]: maybe can make the button smaller

**Review by KirillSerg — APPROVED** [review]

File: package.json (hunk 9 above)
Comment (Mykyta-Pysarev) [reply package.json:14]: ?

File: src/containers/Registration.tsx (hunk 10 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: What is signature? All code formatted with Prettier.

File: src/containers/ModalAuthorization.tsx (hunk 11 above)
Comment (Mykyta-Pysarev) [reply src/containers/ModalAuthorization.tsx]: Why?

File: src/containers/Registration.tsx [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  opacity: 0;
+  padding-right: 0;
+  margin-right: 0;
+  min-width: 125%;
+  transform-origin: 0 0;
+`;
+
+const StyledButton = styled(Button)`
+  margin-top: 1rem;
+  border-radius: 9999px;
```
Comment (Tretmax) [inline src/containers/Registration.tsx]: maybe to use a percentage

File: src/containers/ModalAuthorization.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import { useState } from "react";
+import { Modal, Button } from "react-bootstrap";
+import styled from "styled-components";
+
+import Registration, { windowTitle } from "./Registration";
+
+const StyledModalTitle = styled(Modal.Title)`
+  text-align: center;
+  font-weight: 700;
+  font-size: 2rem;
```
Comment (alexandrtovmach) [inline src/containers/ModalAuthorization.tsx]: are you sure that's not possible to us bootstrap to define this styles?
how about https://getbootstrap.com/docs/4.0/utilities/text/#font-weight-and-italics

File: src/containers/Registration.tsx [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  margin-top: 1rem;
+  padding: 2rem 0.75rem 2rem 1.7rem !important;
+  height: 5.5rem !important;
+  font-size: 1.5rem;
+
+  ${(props) =>
+    props.top &&
+    css`
+      margin-top: 0rem;
+    `}
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: for margin padding you can use bootstrap modificators:
https://getbootstrap.com/docs/4.0/utilities/spacing/

and by using them you don't need to make these conditional styles

File: src/containers/Registration.tsx [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+  label {
+    padding: 2rem 1.7rem;
+  }
+  .form-floating > .form-control:focus ~ label,
+  .form-control:not(:placeholder-shown) ~ label,
+  .form-control:focus ~ label {
+    transform: scale(0.8) translateY(-1.5rem) translateX(0);
+    opacity: 0.65 !important;
+  }
+`;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx:29]: I'm really not sure about this nesting styles, is it important?

File: src/containers/Registration.tsx [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+const StyledCounterLabel = styled.label`
+  text-align: right;
+  opacity: 0;
+  padding-right: 0;
+  margin-right: 0;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: same, can be solved with bootstrap:
https://getbootstrap.com/docs/4.0/utilities/spacing/

File: src/containers/Registration.tsx [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+  min-width: 125%;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: I'm really scared about values like this)) is there a way to do that more elegant?

File: src/containers/Registration.tsx [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+  transform-origin: 0 0;
+`;
+
+const StyledButton = styled(Button)`
+  margin-top: 1rem;
+  border-radius: 9999px;
+  padding: 1rem;
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: bootstrap

File: src/containers/Registration.tsx [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+`;
+
+interface MyForm {
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]:
```suggestion
interface RegistrationForm {
```

File: src/containers/Registration.tsx [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+            type="password"
+            name="confirmPassword"
+            placeholder="Підтвердіть пароль"
+            autoComplete="on"
+          />
+        </StyledFloatingLabel>
+        <p className="text-danger">
+          {errors.confirmPassword && errors?.confirmPassword?.message}
+        </p>
+        <div className="d-grid gap-2">
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]: interesting decision to use grid here, I can't agree with that, but anyway, you should use <Grid> component from bootstrap instead

File: src/containers/ModalAuthorization.tsx [hunk 21]
```diff
@@ -0,0 +1,34 @@
+import { useState } from "react";
```
Comment (alexandrtovmach) [inline src/containers/ModalAuthorization.tsx]: not really what I meant. It wasn't your task, so it's fine and you just improvised, so it's fine to merge as it is, but it's a bit dumb to create <ModalSomething> component for each content that we would put in Modal. Right way to do that is to create reusable and abstract Modal component, that renders Button with text from props, and internally adding onClick handler to this button to open modal with content from props.children.

it's not your task, so doesn't matter

File: src/containers/Registration.tsx [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+          label="Дата народження"
+        >
+          <StyledFormControl
+            {...register("dateOfBirth")}
+            type="date"
+            name="dateOfBirth"
+          />
+        </StyledFloatingLabel>
+        <p className="text-danger">
+          {errors.dateOfBirth && errors?.dateOfBirth?.message}
```
Comment (alexandrtovmach) [inline src/containers/Registration.tsx]:
```suggestion
          {errors?.dateOfBirth?.message}
```

Comment (alexandrtovmach) [reply src/containers/Registration.tsx]: and fix this everywhere

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]
I've resolved all strange comments from other students, please fix only one that I left and added thumbsup
and a few more from me)

File: src/containers/Registration.tsx (hunk 12 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Change from styled component to clean bootstrap. Will keep in mind to use % next time.

File: src/containers/Registration.tsx (hunk 18 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Remade with bootstrap

File: src/containers/Registration.tsx (hunk 19 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+`;
+
+const StyledButton = styled(Button)`
+  margin-top: 1rem;
+  border-radius: 9999px;
+  padding: 1rem;
+`;
+
+interface MyForm {
+  userName: string;
```
Comment (VitaliiDenysiuk) [inline src/containers/Registration.tsx]: in backend field "username"

Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 22 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 20 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: image
This construct is from react bootstrap documentation :)

Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Now using Bootstrap component instead styled component

File: src/containers/Registration.tsx (hunk 16 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Done. Thanks for the tip!

File: src/containers/ModalAuthorization.tsx (hunk 21 above)
Comment (Mykyta-Pysarev) [reply src/containers/ModalAuthorization.tsx]: I think I did what you originally wanted. But need confirmation from you.

File: src/containers/Registration.tsx (hunk 17 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/Registration.tsx (hunk 15 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx:29]: I need this to overwrite original css from bootstrap. With original bootstrap label :focus style is far from perfect (with large font and large input)

File: src/containers/Registration.tsx (hunk 14 above)
Comment (Mykyta-Pysarev) [reply src/containers/Registration.tsx]: Fixed

File: src/containers/ModalAuthorization.tsx (hunk 13 above)
Comment (Mykyta-Pysarev) [reply src/containers/ModalAuthorization.tsx]: Done

**Review by alexandrtovmach — APPROVED** [review]

**Comment by alexandrtovmach** [comment]
fine for now
still not sure about overriding styles for float label inputs, but let's check it later

**Pull request merged**